shared/middleware/files_api.rb

```diff
@@ -486,25 +486,35 @@ def copy_file(endpoint, encrypted_channel_id, filename, source_filename)
   # manifest.
   #
   def files_put_file(encrypted_channel_id, filename, body)
-    downcased_filename = filename.downcase
-    bad_request if downcased_filename == FileBucket::MANIFEST_FILENAME
+    unescaped_filename = CGI.unescape(filename)
+    unescaped_filename_downcased = unescaped_filename.downcase
+    bad_request if unescaped_filename_downcased == FileBucket::MANIFEST_FILENAME
 
     bucket = FileBucket.new
     manifest = get_manifest(bucket, encrypted_channel_id)
     manifest_is_unchanged = true
 
+    unescaped_src_filename_downcased = params['src'] ? CGI.unescape(params['src']).downcase : nil
+    unescaped_delete_filename_downcased = params['delete'] ? CGI.unescape(params['delete']).downcase : nil
+    case_only_rename = unescaped_filename_downcased == unescaped_src_filename_downcased
+
     # store the new file
-    if params['src']
-      new_entry_json = copy_file('files', encrypted_channel_id, downcased_filename, params['src'].downcase)
+    if unescaped_src_filename_downcased
+      if case_only_rename
+        src_entry = manifest.detect {|e| e['filename'].downcase == unescaped_filename_downcased}
+        not_found if src_entry.nil?
+        new_entry_json = src_entry.to_json
+      else
+        new_entry_json = copy_file('files', encrypted_channel_id, URI.encode(unescaped_filename_downcased), URI.encode(unescaped_src_filename_downcased))
```

Comment:
nit: This line is a too long, please break it up as
new_entry_json = copy_file(
  'files',
  encrypted_channel_id,
  URI.encode(unescaped_filename_downcased),
  URI.encode(unescaped_src_filename_downcased)
)
```